### DISCUSSION TITLE
storage and state management on gno.land

### DISCUSSION THREAD

**Discussion by felipebrunet** [gnolang/gno#2845] [General]
Hi
Context: I wonder how will data be managed in Gno blockchains.
Question1:
I assume it is through transactions that you can edit/add data (changing the state of realms). However I can also edit those files directly in my terminal as a node operator of a local gnoland. What will happen if one node operator edits data directly? Will that node be excluded from consensus?

Question 2:
The amount of data a gnoland blockchain can have is always incremental or it can decrease with transactions? For example a function "Delete_post" will actually delete text in a file?

Question 3:
As Gnoland blockchain grow in realms and blocks, how do you think data will be handled? Will it be through pruning older states?
Will it be through sharding? 
If that is the case, are those parallel chains going to have a separate GNOT coin or will those chains have the same coin of the original Gnoland chain? I suppose the latter will increase value of the original coin, but, is it possible to start a side gnoland chain without its own coin?

**Comment by thehowl** [comment]
hey. 

> I assume it is through transactions that you can edit/add data (changing the state of realms). However I can also edit those files directly in my terminal as a node operator of a local gnoland. What will happen if one node operator edits data directly? Will that node be excluded from consensus?

gnodev is a local development enviroment; which is different than what happens on-chain. the files are "modified" by actually re-creating the blockchain with a new genesis file, containing the new, updated files. IOW, a node operator cannot "directly" edit files.

do note that node operators can do all kinds of things to deviate from the "consensus". when that happens, though, their nodes generally won't sync with the rest. we haven't figured out all the details yet, but in cases of malicious nodes, we'll have mechanisms to vote them out of the valset and/or slash them.

> The amount of data a gnoland blockchain can have is always incremental or it can decrease with transactions? For example a function "Delete_post" will actually delete text in a file?

the state can grow and shrink based on the operations being perfomed (the transactions). the transaction history is linearly growing; however a node can eventually prune the transaction history. there will be specialized "archive nodes" to keep history up to the beginning of the chain.

>As Gnoland blockchain grow in realms and blocks, how do you think data will be handled? Will it be through pruning older states?

this is a complex question, where the answer is still being developed and in flux, so i can't give you a straight answer yet.

a few things to note:

- gno.land follows in the cosmos philosophy of not building "everything" onto one chain; but rather being able to develop more specialised chains for different use cases. this will likely not be the case for a few years after launch: we'll need to continue work on polishing the vm and the specifications for many things, so alternative VMs can also be built on the same programming language. but the "horizontal scaling" will eventually be through this.
- we're likely to have some kind of mechanism to cap the space available to contracts; and then increase the space available in the state gradually.
- the two main usecases we're building gno.land for, for now: (1) building a hub of good, foundational packages and realms, upon which others can be built, and (2) building social applications, similar to twitter/reddit/etc., to discuss with other people in a decentralized manner and hopefully build "saner" spaces than these.

**Reply by felipebrunet** [reply]
Hey thanks for the thorough answer!!

Yes my last question about scaling was putting as example (in my mind) reddit, twitter and Wikipedia. If its a lot of data, having many chains storing it in parts would mean less data to be stored by each node.

Cheers!
